### PR TITLE
Upload pretrained models

### LINKED ISSUE #106: [FEATURE] add support for MLCommons Pre-trained models support
**Is your feature request related to a problem?**

No.

From Opensearch 2.6 version, MLCommons support [Pre-trained models](https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/).

But from opensearch-py-ml's [upload_model](https://github.com/opensearch-project/opensearch-py-ml/blob/main/opensearch_py_ml/ml_commons/ml_commons_client.py#L28-L33) doesn't support that yet. 

Either we need to add another function `upload_pretrained_model` or include uploading pre-trained model feature in `upload_model` method

### PR DESCRIPTION
**Pull request by AlibiZhenis**
### Description
Added a feature of uploading pretrained models. Closes #106 .
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 1]
```diff
@@ -63,6 +63,71 @@ def upload_model(
             model_path, model_config_path, isVerbose
         )
 
+    def upload_pretrained_model(self, name: str, version: str, model_format):
```
[the remark this answers is not in the document]
Comment (AlibiZhenis) [reply opensearch_py_ml/ml_commons/ml_commons_client.py:66]: So, if I'm not mistaken, to upload a pretrained model, we have to:

1. Send model info to /models/_upload
2. Then load the model using load model api, after which ML Commons downloads the model chunks from Hugging Face by itself.

If I do the above 2 steps, I can successfully use the model for inference in dev tools.

Are you saying that that I shouldn't do the second part? Because without it, if I try to do inference, I get the "model not loaded" error. 

I would appreciate it if you clarified this for me.

[the remark this answers is not in the document]
Comment (AlibiZhenis) [reply opensearch_py_ml/ml_commons/ml_commons_client.py]: I think it's a great idea. The argument was added.

Also, about sending model info to upload api, that's what _create_model_metadoc method does anyway. I just misnamed the method.

Both of these issues were fixed in the last commit.

**Comment by dhrubo-os** [comment]
Thanks for the PR. It looks good. 

Integration test is failing. I believe this is because right now our integration test is based on [opensearch 2.5.0](https://github.com/opensearch-project/opensearch-py-ml/blob/main/.github/workflows/integration.yml#L15)

But this new feature is in `opensearch 2.6.0`. So we should update that version to pass the integration test.

Lint check is also failing. [Check here](https://github.com/opensearch-project/opensearch-py-ml/blob/main/DEVELOPER_GUIDE.md#to-test-code-formatting-and-linting-issue-we-can-run)

**Comment by AlibiZhenis** [comment]
So what should I do next? I have opensearch-py 2.2.0 installed, which pip says is the latest version.

**Comment by dhrubo-os** [comment]
> So what should I do next? I have opensearch-py 2.2.0 installed, which pip says is the latest version.



> 

Can you please change to `2.6.0` [here])https://github.com/opensearch-project/opensearch-py-ml/blob/main/.github/workflows/integration.yml#L15) ?

**Comment by dhrubo-os** [comment]
You can also run the whole test in your end like this: `./.ci/run-tests "opensearch" "true" 2.6.0` to check

**Comment by AlibiZhenis** [comment]
It seems the test was failing because model unload was timing out. Once I increased the threshold to 60s, it's now passing on my side.

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 2]
```diff
@@ -63,6 +63,76 @@ def upload_model(
             model_path, model_config_path, isVerbose
         )
 
+    def upload_pretrained_model(
+        self, name: str, version: str, model_format: str, load_model: bool = True
+    ):
+        """
+        This method uploads a pretrained model into opensearch cluster using ml-common plugin's api.
+        First, this method creates a model id to store model info
+        and then loads the model from Hugging Face if load_model is True.
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_commons/ml_commons_client.py:72]: then loads the model in memory. We don't load model from huggingFace. Model is downloaded in their index and then it will be loaded into memory using DJL.

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 3]
```diff
@@ -63,6 +63,76 @@ def upload_model(
             model_path, model_config_path, isVerbose
         )
 
+    def upload_pretrained_model(
+        self, name: str, version: str, model_format: str, load_model: bool = True
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_commons/ml_commons_client.py:67]: how about `model_name`, `model_version`?

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 4]
```diff
@@ -63,6 +63,76 @@ def upload_model(
...
+        This method uploads a pretrained model into opensearch cluster using ml-common plugin's api.
+        First, this method creates a model id to store model info
+        and then loads the model from Hugging Face if load_model is True.
+        The model has to be supported by ML Commons. Refer to https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/.
+
+        :param name: Name of the pretrained model
+        :type name: string
+        :param version: Version of the pretrained model
+        :type version: string
+        :param model_format: Currently only supports "TORCH_SCRIPT"
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_commons/ml_commons_client.py:79]: We also support `ONNX`. If you see this : https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/, we also have urls for `ONNX` artifacts.

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 5]
```diff
@@ -63,6 +63,76 @@ def upload_model(
...
+        :type model_format: string
+        :param load_model: Whether to load the model chunks from Hugging Face
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_commons/ml_commons_client.py:81]: whether to load the model in memory using uploaded model chunks.

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 6]
```diff
@@ -63,6 +63,76 @@ def upload_model(
...
+                raise Exception(output["error"])
+
+        return model_id
+
+    def _send_model_info(self, model_meta_json: dict):
+        """
+        This method sends the pretrained model info to ML Commons' upload api
+
+        :param model_meta_json: a dictionary object with model configurations
+        :type model_meta_json: string
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_commons/ml_commons_client.py:113]: dict instead of string

File: opensearch_py_ml/ml_commons/ml_commons_client.py [hunk 7]
```diff
@@ -63,6 +63,76 @@ def upload_model(
...
+        :return: returns a unique id of the model
+        :rtype: int
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_commons/ml_commons_client.py:115]: string instead of int

**Comment by dhrubo-os** [comment]
When you push a new commit with all these changes can you also run the workflow. Please let me know if you don't have permission to run workflow?

**Comment by AlibiZhenis** [comment]
Sorry for all the mistakes. I'll try to be more attentive from now on.

It also seems that I don't have permission to run workflows.

File: tests/ml_commons/test_ml_commons_client.py [hunk 8]
```diff
@@ -70,6 +74,47 @@ def test_init():
...
+        assert raised == False, "Raised Exception in getting pretrained model info"
+
+        raised = False
+        try:
+            ml_client.unload_model(model_id)
+            time.sleep(60)
+            ml_model_status = ml_client.get_model_info(model_id)
+            print("ml_model_status", ml_model_status)
+            assert ml_model_status.get("model_state") == "UNLOADED"
+        except:
```
Comment (dhrubo-os) [inline tests/ml_commons/test_ml_commons_client.py:106]: Lint is failing for bare except, you need to add to skip linting error `# noqa: E722`

File: tests/ml_commons/test_ml_commons_client.py [hunk 9]
```diff
@@ -70,6 +74,47 @@ def test_init():
...
+            assert ml_model_status.get("model_format") == "TORCH_SCRIPT"
+            assert ml_model_status.get("algorithm") == "TEXT_EMBEDDING"
+        except:  # noqa: E722
+            raised = True
+        assert raised == False, "Raised Exception in getting pretrained model info"
+
+        raised = False
+        try:
+            ml_client.unload_model(model_id)
+            time.sleep(60)
```
Comment (dhrubo-os) [inline tests/ml_commons/test_ml_commons_client.py:102]: Integration is still failing. May be we can have a loop to check back in every 15 seconds. if we get the result, that's great otherwise will wait for another 15 seconds. 

But we can't run this forever. May be the hard limit can be 5 minutes. IF total duration passes more than 5 minutes then fail the test.

**Comment by dhrubo-os** [comment]
> > So what should I do next? I have opensearch-py 2.2.0 installed, which pip says is the latest version.
> 
> Can you please change to `2.6.0` [here](https://github.com/opensearch-project/opensearch-py-ml/blob/main/.github/workflows/integration.yml#L15) ?

I saw you changed the version, but now i don't see the changed code. Where is it? Integration test is failing....

File: .github/workflows/integration.yml [hunk 10]
```diff
@@ -12,7 +12,7 @@ jobs:
         cluster: ["opensearch"]
         secured: ["true"]
         entry:
-          - { opensearch_version: 2.6.0 }
+          - { opensearch_version: 2.5.0 }
```
Comment (dhrubo-os) [inline .github/workflows/integration.yml]: Why did you change this back?

Comment (AlibiZhenis) [reply .github/workflows/integration.yml]: The tests are passing in my local environment, but not the workflow. So I wanted to see if the problem was with the version. Turns out it's not, I will change it back.

I am suspecting the problem is with the circuit breaker error. I am investigating.

**Comment by AlibiZhenis** [comment]
It seems to be consistently passing now once I changed the version back to 2.6.0. I believe the reason is in my "minor code and test changes", where I changed the upload_pretrained_model method.

Now the method finishes once the model_state becomes loaded, instead of when the load model task gets completed.

**Review by dhrubo-os — APPROVED** [review]

**Review by b4sjoo — APPROVED** [review]
